### PR TITLE
[SYCL] Fix command cleanup invoked from multiple threads

### PR DESCRIPTION
**Pull request by sergey-semenov**
This patch fixes a sporadic bug where one thread attempted to
clean up a command already deleted by another.

Signed-off-by: Sergey Semenov <sergey.semenov@intel.com>

### CODE REVIEW

File: sycl/source/detail/scheduler/scheduler.cpp [hunk 1]
```diff
@@ -123,9 +123,11 @@ void Scheduler::waitForEvent(EventImplPtr Event) {
   GraphProcessor::waitForEvent(std::move(Event));
 }
 
-void Scheduler::cleanupFinishedCommands(Command *FinishedCmd) {
+void Scheduler::cleanupFinishedCommands(EventImplPtr FinishedEvent) {
   std::lock_guard<std::mutex> lock(MGraphLock);
-  MGraphBuilder.cleanupFinishedCommands(FinishedCmd);
+  Command *FinishedCmd = static_cast<Command *>(FinishedEvent->getCommand());
+  if (FinishedCmd)
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/scheduler.cpp]: Please add a comment why we check here again.

File: sycl/test/scheduler/CommandCleanupThreadSafety.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#include <cassert>
+#include <cstddef>
+#include <thread>
+#include <vector>
+
+// This test checks that the command graph cleanup works properly when
+// invoked from multiple threads.
+using namespace cl::sycl;
+
+class Foo;
```
Comment (romanovvlad) [inline sycl/test/scheduler/CommandCleanupThreadSafety.cpp]: Probably it should be better to move to ./unittests/thread_safety for consistency. But I'm OK with LIT test.

Comment (sergey-semenov) [reply sycl/test/scheduler/CommandCleanupThreadSafety.cpp]: This will be addressed later

Comment (romanovvlad) [reply sycl/test/scheduler/CommandCleanupThreadSafety.cpp]: Ok

**Pull request merged**